sycl/source/detail/platform_impl.cpp

```diff
@@ -196,6 +196,40 @@ static void filterAllowList(vector_class<RT::PiDevice> &PiDevices,
   PiDevices.resize(InsertIDx);
 }
 
+// @return True if the device is invalid for the current backend preferences
+static bool isDeviceInvalidForBe(const device &Device) {
+
+  if (Device.is_host())
+    return false;
+
+  // Retrieve Platform version to identify CUDA OpenCL platform
+  // String:  OpenCL 1.2 CUDA <version>
+  const platform platform = Device.get_info<info::device::platform>();
+  const std::string platformVersion =
+      platform.get_info<info::platform::version>();
+  const bool HasOpenCL = (platformVersion.find("OpenCL") != std::string::npos);
+  const bool HasCUDA = (platformVersion.find("CUDA") != std::string::npos);
+
+  backend *PrefBackend = detail::SYCLConfig<detail::SYCL_BE>::get();
+  auto DeviceBackend = detail::getSyclObjImpl(Device)->getPlugin().getBackend();
+
+  // Reject the NVIDIA OpenCL implementation
+  if (DeviceBackend == backend::opencl && HasCUDA && HasOpenCL)
+    return true;
+
+  // If no preference, assume OpenCL and reject CUDA
+  if (DeviceBackend == backend::cuda && !PrefBackend) {
+    return true;
+  } else if (!PrefBackend)
+    return false;
+
+  // If using PI_OPENCL, reject the CUDA backend
+  if (DeviceBackend == backend::cuda && *PrefBackend == backend::opencl)
+    return true;
+
+  return false;
+}
```

Comment:
```suggestion
  // If no preference, assume OpenCL and reject CUDA
  if (DeviceBackend == backend::cuda && (!PrefBackend || *PrefBackend == backend::opencl))
    return true;

  return false;
}
```

Reply:
> Minor fixes to CUDA backend to get BabelStream (https://github.com/UoB-HPC/BabelStream), each one on individual commits:
> 
> * Fixed: default selection for CUDA devices in presence of the NVIDIA OpenCL platform
> * Fixed: Missing return event on Unmap
> * Throw error in queue constructor when backend is not correct (currently segfaults)

From the PR description, this piece of code seems to be "not required to run BabelStream benchmarks on CUDA" and removing NVIDIA OpenCL implementation is enough. Right?

If so, @Ruyk, can we move this part to a separate PR to make @hiaselhans happy?

Reply:
Sure, i'll split out the changes

Reply:
> not required to run BabelStream benchmarks on CUDA" and removing NVIDIA OpenCL > implementation is enough. Right?

Actually, babel stream doesnt use a device selector but a device list, so the NVIDIA OpenCL still appears. There are other workarounds I can do so not a problem.

Reply:
> If so, @Ruyk, can we move this part to a separate PR to make @hiaselhans happy?

thanks @bader :smile: 
well, maybe it's me being a dickhead with this.
https://github.com/intel/llvm/blob/cf65794dd6352e9d3f8fd47665e3be5e86a9c7fb/sycl/source/device.cpp#L54
just saying we're re-implementing things already implemented somewhere else...